neutron_policy_server/wsgi.py

```diff
@@ -16,40 +19,85 @@
-                    mac_address=allowed_address_pair["mac_address"],
-                )
-
-            if len(ports) != 1:
-                return Response(status=403)
-
-            fixed_ips = [str(fixed_ip["ip_address"]) for fixed_ip in ports[0].fixed_ips]
-            if allowed_address_pair["ip_address"] not in fixed_ips:
-                return Response(status=403)
-
-        return Response(status=200)
+            ports = port_obj.Port.get_objects(
+                ctx,
+                network_id=target["network_id"],
+                project_id=target["project_id"],
+                mac_address=allowed_address_pair["mac_address"],
+            )
+        if len(ports) != 1:
+            return Response("Zero or Multiple match port found.", status=403, mimetype="text/plain")
+        fixed_ips = [str(fixed_ip["ip_address"]) for fixed_ip in ports[0].fixed_ips]
+        if allowed_address_pair["ip_address"] not in fixed_ips:
+            return Response("IP address not exists in ports.", status=403, mimetype="text/plain")
+    return Response("True", status=200, mimetype="text/plain")
+
+
+@app.route("/port-update", methods=["POST"])
+def enforce_port_update():
+    target, ctx = _fetch_context(request)
+    if "mac_address" not in target and "fixed_ips" not in target:
+        return Response("True", status=200, mimetype="text/plain")
+    with db_api.CONTEXT_READER.using(ctx):
+        ports = port_obj.Port.get_objects(ctx, id=target["id"])
+        if len(ports) == 0:
+            return Response("No match port found.", status=403, mimetype="text/plain")
+
+        fixed_ips = [str(fixed_ip["ip_address"]) for fixed_ip in ports[0].fixed_ips]
+
+        query = ctx.session.query(models.AllowedAddressPair).filter(
+            models.AllowedAddressPair.mac_address.in_([ports[0].mac_address])
+        ).filter(models.AllowedAddressPair.ip_address.in_(fixed_ips))
+
+    pairs = [aap_obj.AllowedAddressPair._load_object(context, db_obj)
+    for db_obj in query.all()]
+    if len(pairs) > 0:
+        return Response("Address pairs dependency found for this port.", status=403, mimetype="text/plain")
+    return Response("True", status=200, mimetype="text/plain")
+
+
+@app.route("/port-delete", methods=["POST"])
+def enforce_port_delete():
+    target, ctx = _fetch_context(request)
+
+    fixed_ips = [str(fixed_ip["ip_address"]) for fixed_ip in target["fixed_ips"]]
+    with db_api.CONTEXT_READER.using(ctx):
+        query = ctx.session.query(models.AllowedAddressPair).filter(
+            models.AllowedAddressPair.mac_address.in_([target["mac_address"]])
+        ).filter(models.AllowedAddressPair.ip_address.in_(fixed_ips))
+
+    pairs = [aap_obj.AllowedAddressPair._load_object(context, db_obj)
+    for db_obj in query.all()]
+    if len(pairs) > 0:
+        return Response("Address pairs dependency found for this port.", status=403, mimetype="text/plain")
+    return Response("True", status=200, mimetype="text/plain")
```

Comment:
any reason why we can't use the `rule` and have a single entry point for the whole policy server?

Reply:
the oslo.policy http check from neutron didn't provide rule information from `neutron/pecan_wsgi/hooks/policy_enforcement.py` at least that's what my new test environment shows.
will check in more detail from code to see if that's a neutron bug or just my environment issue.

Reply:
It's a bug in oslo.policy -> HttpCheck

Reply:
https://review.opendev.org/c/openstack/oslo.policy/+/921913

Reply:
Oh that sucks, since we're going to have to handle older version, I see why you've broken it down.  Stinky.